### ISSUE TITLE
Slowdown with slicing involving large 1D index arrays

### ISSUE THREAD

**Issue by andrewcollette** [h5py/h5py#293]
From https://groups.google.com/forum/?fromgroups=#!topic/h5py/gpo19B-xGMc

Evidently using 1D index arrays on the order of 20k points results in a slowdown, whereas 20 discrete 1k arrays work OK.  May be related to the slicing code in `h5py._hl.selections`.  cProfile suggests most of our time is spent in select_hyperslab.

**Comment by oruebel** [comment]
We have come across this problem also here https://github.com/NeurodataWithoutBorders/pynwb/issues/1276 and I was wondering whether a) this is being addressed in h5py/HDF5 or b) whether there is a best-practice solution for this?

**Comment by rvinas** [comment]
Is there any update on this?

**Comment by aaroncnb** [comment]
**Curious if any updates have occurred on this issue?**
(cc: [@timothygebhard](https://github.com/timothygebhard/complexipy))

We are encountering the same issue, as part of our [Frontier Development Lab 2022 ](https://frontierdevelopmentlab.org/)"research incubator" sprint, on the Astrobiology Team.

No `Exceptions` are thrown, but the process slows down and becomes unresponsive to the point that we cannot even do a `KeyboardInterrupt`

Happy to provide further diagnostic info!
We're just not sure what to provide at the moment, since no error messages appear.

Thanks in advance for any workarounds, updates, or suggested code for us to look at and aid in the debugging effort!!

I'll attach the relevant function as a text file below!

TLDR: `np.array(hdf_file[feature][idx])` is :

- slow**er** for small idx but always works
- prohibitively slow if idx is too long

[data.txt](https://github.com/h5py/h5py/files/9204282/data.txt)

**Comment by tacaswell** [comment]
> ... becomes unresponsive to the point that we cannot even do a KeyboardInterrupt

This suggests that the slow down is in the c-layer (the signal handler belongs to the Python main loop and sets a flag to raise `KeyboardInterupt` on the next byte code to be interpreted.  If it is not happening that means that all of the time is in a single byte code evaluation -> we down in the c layer.

I think the next step would be to add some prints in the cython code to check if it is loops we are doing in cython or it is slowing down in libhdf5.

It may also be worth identifying these inputs and seeing if they can be transformed to a selection that is faster.

**Comment by oruebel** [comment]
Has there been any progress on this issue?

**Comment by cwindolf** [comment]
One very common way that I end up hitting this issue is when doing random access along the first dimension of a dataset which is only chunked along the first dimension. I think that pattern is pretty common in applications where samples are indexed along the first dimension.

In this case, it's possible to work around this issue using `Dataset.iter_chunks`. I use the following helper function to implement the workaround:

```python
def _read_by_chunk(mask, dataset):
    """
    mask : boolean array of shape dataset.shape[:1]
    dataset : chunked h5py.Dataset
    """
    out = np.empty((mask.sum(), *dataset.shape[1:]), dtype=dataset.dtype)
    n = 0
    for sli, *_ in dataset.iter_chunks():
        m = np.flatnonzero(mask[sli])
        nm = m.size
        if not nm:
            continue
        x = dataset[sli][m]
        out[n : n + nm] = x
        n += nm
    return out
```

Here, `mask` is a Boolean array indicating which samples to grab. Below are some benchmarks on a real world application, coincidentally on some neuroscience data, which may be relevant to the NWB issue mentioned above (https://github.com/NeurodataWithoutBorders/pynwb/issues/1276).

Setup and info:
```
print(mask.shape, mask.sum(), mask.mean())
# prints: (1971390,), 211510, 0.10728978030729587

indices = np.flatnonzero(mask)

with h5py.File(h5_path, "r", locking=False) as h5:
    print(h5[dataset_name].shape)
    print(h5[dataset_name].chunks)
# prints: (1971390, 8, 198)
# prints: (1024, 8, 198)
```

Timing `_read_by_chunk`:
```
%%timeit
with h5py.File(h5_path, "r") as h5:
    x = _read_by_chunk(mask, h5[dataset_name])
# prints: 791 ms ± 2.55 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Timing reading full dataset into memory
```
%%timeit
with h5py.File(h5_path, "r") as h5:
    x = h5[dataset_name][:][indices]
# prints: 4.77 s ± 8.01 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Timing against reading 128 indices at a time:
```
%%timeit
with h5py.File(h5_path, "r", locking=False) as h5:
    out = np.empty((indices.size, *h5[dataset_name].shape[1:]), dtype=h5[dataset_name].dtype)
    for i in range(0, len(indices), 128):
        out[i:i+128] = h5[dataset_name][indices[i:i+128]]
# prints: 23.7 s ± 232 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

And, I also tried benchmarking the strategy of letting HDF5 do the work (`x = h5[dataset_name][indices]`), but it did not complete after >10 minutes and I didn't have more time to wait.

In this benchmark at least, `_read_by_chunk` is pretty fast, and much better than competing strategies. The dataset is float32, so we're reading ~1.3GB of data in `mask`. I'm not too sure how to control the ways that these benchmarks interact with file system caches etc, so please forgive any issues there (and let me know if I've made mistakes).

It would be awesome if we could read from h5s without such workarounds -- I would in theory be interested in helping with any possible PRs along those lines, but I took a look at the code in `_hl/selections.py` and was a bit daunted :). I'm also not sure how this strategy would interact with non-chunked storage or data which is chunked along multiple dimensions. Still .h5 is my favorite resizeable and portable array format, and I wish I had had this code when I first saw this github issue a few years ago, so I thought I'd share it for future visitors :).

**Comment by tacaswell** [comment]
Very through analysis @cwindolf !  I think this makes sense and suggests that something in either how we are calling libhdf5 or in libhdf5 itself is going sideways.  If I had to make a wild guess it would be either there is a line-painter bug in allocating the output array (turning a linear process into a quadratic one) or it is doing IO for every entry.

https://github.com/h5py/h5py/blob/c8972f1a93b9b55620e617a615d3a912a71bff05/h5py/_hl/dataset.py#L830-L831 looks to be where in `__getitem__` we create the selector objects and https://github.com/h5py/h5py/blob/c8972f1a93b9b55620e617a615d3a912a71bff05/h5py/_hl/selections.py#L62-L65 is there we create the `PointSelection` object to use.

I suspect the right path is to add special casing to `__getitem__` for big point selections (either before we create the `PointSelection` object or after) and embed a version of `_read_by_chunk`  that can handle multiple-dimension chunking (or really restrict the special case).

Writing an example c++ program that using `PointSelection` and shows the same pathologically bad run time would also be useful (and help sort out if this is really an h5py issue or a libhdf5 issue).

**Comment by ilan-gold** [comment]
Commenting as much because I have a nice repro as a I wish to be made aware once this is resolved:

```python
from timeit import timeit

import h5py
import numpy as np

N = 100_000

with h5py.File("test1.h5", "w") as f1:
    f1["data"] = np.arange(N)
    key = np.random.randint(0, f1["data"].shape[0], (f1["data"].shape[0] // 2,))
    key.sort()
    key = np.unique(key)
    print(f"fetching from disk {len(key)} points from {f1['data']}")
    print(timeit(lambda: f1["data"][np.unique(key)], number=1))
    print(f"fetching from memory {len(key)} points from {f1['data']}")
    print(timeit(lambda: f1["data"][...][np.unique(key)], number=1))
```
Thanks!
